cde/doc/cde1/man/man1/dtappgather.1
.\"----------------------------------------------------------------------------
.\"----------------------------------------------------------------------------
.TH dtappgather 1 "22 August 1994"
.BH "22 August -1994"
.\"----------------------------------------------------------------------------
.\"---Define variables used by this man page
.de V1
\fBdtappgather\fR
..
.de V2
Application Manager
..
.de V3
\fBdtsearchpath\fR
..
.de V4
\f(CWReloadApps\fR
..
.de V5
\f(CWDTAPPSEARCHPATH\fR
..
.\"---
.\"----------------------------------------------------------------------------
.\" (c) Copyright 1993, 1994 Hewlett-Packard Company
.\" (c) Copyright 1993, 1994 International Business Machines Corp.
.\" (c) Copyright 1993, 1994 Sun Microsystems, Inc.
.\" (c) Copyright 1993, 1994 Novell, Inc.
.SH NAME
dtappgather \- Gather application files for presentation by the 
Application Manager.
.\"----------------------------------------------------------------------------
.SH SYNOPSIS
.V1
[ \fB-r\fR ]
.SH DESCRIPTION
The 
.V2
presents personal, system, and factory applications in
an easy-to-access window.
The
.V2
is implemented as a special File Manager view of a 
subdirectory that is built on a per-user, per-session basis.
The
.V1
utility is responsible for creating and refreshing
the user's
.V2 
subdirectory.
.PP
.V1 
is run at login and whenever the user invokes the
.V4
action.
The user's individual subdirectory for the 
Application Manager, \f(CW/var/dt/appconfig/appmanager/$DTUSERSESSION\fR,
is opened when the user presses the Application Manager control
on the Front Panel.
.P
The sequence of events is as follows:
.TP 4
\(em
When the user logs in, the \fBXsession\fR script sources in the
Xsession.d script that sets the \f(CWDTUSERSESSION\fR environment
variable.
.TP 4
\(em
\fBXsession\fR then invokes the
.V3
utility to set the \f(CWDTAPPSEARCHPATH\fR environment variable.
.RE
.TP 4
\(em
After 
.V3
returns with the \f(CWDTAPPSEARCHPATH\fR assembled,
the \fBXsession\fR script calls \fBdtappgather\fR.
.TP 4
\(em
As its main function,
.V1
traverses the \f(CWDTAPPSEARCHPATH\fR,
examining each possible source of applications and, where
there are existing source subdirectories, creates symbolic links between
the source and the user's \f(CWDTUSERSESSION\fR subdirectory.
.TP 4
\(em
Finally,
.V1
turns off write permissions on the resulting subdirectory to ensure its
integrity.
.PP
A similar sequence occurs when the user double-clicks the \f(CWReloadApps\fR
action after logging in.
.PP
Although the value of the \f(CWDTAPPSEARCHPATH\fR can be modified in a
local shell, the
.V2
view based on its value is \fInot\fR functional until you log out and
log back in. Since this can be a tedious venture, you can verify the
.V2
view by executing dtsearchpath and dtappgather in the local shell. For
example, to add host trout as a system-wide
application server, set the dtsearchpath input environment variable,
DTSPSYSAPPHOSTS:
.P
.in +5
DTSPSYSAPPHOSTS=trout:
.in -5
.P
Then, execute dtsearchpath to update the DTAPPSEARCHPATH environment
variable.
.P
.in +5
eval `dtsearchpath`
.in -5
.P
Finally, gather the new applications by executing dtappgather. The
.V2
will show the new application groups but will not be functional.
.\"----------------------------------------------------------------------------
.SH OPTIONS
.TP
\fB-r\fR
Causes
.V1
to retain the previous contents of the Application Manager, although
discarding broken links.  At login,
.V1
destroys the user's previous \f(CWDTUSERSESSION\fR subdirectory before
creating a new one.  At \f(CWReloadApps\fR time, the \fB-r\fR option
is used to minimize visual disruption of any opened
.V2
views.
.SH RETURN VALUES
.TP 15
0
The command always returns 0 for successful completion.
.SH ENVIRONMENT
.LP
.TP 15
.SM DTAPPSEARCHPATH
Set by the 
.V3
utility.  \f(CWDTAPPSEARCHPATH\fR controls the places
where 
.V1 
will gather applications.  The default locations consist of
\f(CW$HOME/.dt/appmanager\fR (for end users),
\f(CW/etc/dt/appconfig/appmanager/$LANG\fR (for system administrators), and
\f(CW/usr/dt/appconfig/appmanager/$LANG\fR (for factory applications).
.TP 15
.SM DTUSERSESSION
Controls the location of the end user's subdirectory where the
Application Manager will be rooted.  The subdirectory name includes
both the user's \f(CW$LOGNAME\fR and \f(CW$DISPLAY\fR in order to
ensure the user's view of the Application Manager remains consistent
across sessions.
.SH LOCALES
When searching for system administrator and factory default 
.V2
files,
.V1
always uses a value of \f(CW$LANG\fR equal to \f(CWC\fR.  When the
user picks a language at login, the corresponding \f(CW$LANG\fR
subdirectories are searched.  If multiple $LANG subdirectories exist
in \f(CW/etc/dt/appconfig/appmanager\fR, all $LANG subdirectories will
appear in the Application Manager. However, if multiple $LANG
subdirectories exist in /usr/dt/appconfig/appmanager then the
language-specific subdirectory based on the current value of $LANG is
chosen instead of the \f(CWC\fR subdirectory.
.P
Regardless of locale, all the action files that exist under the user's
\f(CW$HOME/.dt/appmanager\fR subdirectory will be symbolically linked
to the user's
.V2
subdirectory.
.SH FILES
.PD 0
.TP 20
.B /usr/dt/bin/Xsession
Among its tasks at login,
the \fBXsession\fR script invokes 
.V3
and then
\fBdtappgather\fR.
.TP 20
.B /var/dt/appconfig/appmanager/$DTUSERSESSION
This subdirectory is where
.V1
assembles the 
.V2
view for the particular user and CDE session.
.PD
.SH "SEE ALSO"
\fBdtappintegrate(1)\fR,
\fBdtsearchpath(1)\fR
.SH NOTES
.LP
In the case of multiple search locations having the \fIsame name\fR,
.V1
picks only the first and discards the rest.  For example, if the
system administrator sets
.PP
.RS 4
\f(CWDTSPSYSAPPHOSTS=tuna:,trout:\fR
.RE
.PP
and if both hosts \f(CWtuna\fR and \f(CWtrout\fR have an
\f(CW/etc/dt/appconfig/appmanager/C/Editors\fR folder, then only the
\f(CWEditors\fR folder from \f(CWtuna\fR (the first host) will appear
in the user's Application Manager.
.PP
After building the user's \f(CWDTUSERSESSION\fR,
.V1
turns off write permissions on that subdirectory to disallow alteration
by the end user.  Although the end user can resize the
window and rearrange the icons within the window, the Application
Manager is intended to be a read-only source of local
and networked applications.
.\""""""""""""""""""""""""""""eof


cde/cde1/dtappgather/dtappgather.C
/*******************************************************************
**  (c) Copyright Hewlett-Packard Company, 1990, 1991, 1992, 1993.
**  All rights are reserved.  Copying or other reproduction of this
**  program except for archival purposes is prohibited without prior
**  written consent of Hewlett-Packard Company.
********************************************************************
****************************<+>*************************************/
/*******************************************************************
 *
 *  File:              dtappgather.C
 *
 *  Purpose:           Create the Application Manager
 *                     the desktop components.
 *
 *  Description:       dtappgather is invoked by the Xsession file at
 *                     user login.  It collects the application groups
 *                     and subdirectories that are to appear in the
 *                     user's Application Manager view.
 *
 *                     It is also invoked by the "ReloadApps" action.
 *
 *  Product:           @(#)Common Desktop Environment 1.0          
 *
 *  Revision:          $XConsortium: dtappgather.C /main/cde1_maint/1 1995/07/14 22:59:13 drk $
 *
 ********************************************************************/

/********************************************************************
 *
 *  Environment variables used by this utility
 *
 *  DTMOUNTPOINT
 *     Default may be vendor specific:  /nfs
 *
 *  DTAPPSEARCHPATH
 *     Default:  $HOME/.dt,/etc/dt/appconfig,/usr/dt/appconfig
 *     Set at login by the dtsearchpath utility.
 *
 *  HOME
 *     No default.
 *     Set by the dtlogin program.
 *
 *  DISPLAY
 *     No default.
 *     Set by the dtlogin program.
 *
 *  TARGET_APPMAN_DIR
 *     Default:	/var/dt/appconfig/appmanager/$DTUSERSESSION
 *     Set within this utility.
 *
 ****************************************************************************/
#include "Options.h"
#include "dtappgather.h"
#include "DirIterator.h"
#include <stdlib.h>
#include <iostream.h>

Options * options;

void LegacyCleanUp
	(
	CDEEnvironment * user
	)
{

    // For now, remove the old ApplicationManager subdirectory
    // so that we don't leave old CDE users with extra baggage
    // under $HOME.

    CString targetAppmanDir (user->HOME() + ".dt/ApplicationManager");

    if (user->OS()->isDirectory(targetAppmanDir)) {
	user->OS()->changePermissions(targetAppmanDir,0777);
	user->OS()->removeDirectory(targetAppmanDir);
    }
}

AppManagerDirectory::AppManagerDirectory
	(
	CDEEnvironment * user,
	const CString &  app
	) : user_(user),
	    langVersionFound(0),
	    appsp_(app)
{
    // Set the users Application Manager subdirectory

    CString userhostdir = user->UserHostDir();

    if (userhostdir.isNull())
	userhostdir = "generic-display-0";

    dirname_ = "/var/dt/appconfig/appmanager/";
    dirname_ += userhostdir;

    if (user->OS()->isDirectory(dirname_)) {
	user->OS()->changePermissions(dirname_,0755);
	user->OS()->removeDeadLinks(dirname_);
	if (!options->Retain())
	    user->OS()->removeFiles(dirname_,"[.]*");
    }
    else {

	// Make /var/dt/appconfig/appmanger directories if not present 
        //
	// Make /var/dt/tmp directory if not present 
	//
        // Make user session subdirectory under /var/dt/appconfig/appmanager

	CString dir(dirname_);
	dir.replace("/" + userhostdir,"");
	if (!user->OS()->isDirectory(dir)) {	// does appmanager exist?
	    dir.replace("/appmanager","");
	    if (!user->OS()->isDirectory(dir)) {   // does appconfig exist?
		dir.replace("/appconfig","");
	        if (!user->OS()->isDirectory(dir)) {   // does dt exist?
		    dir.replace("/dt","");
		    if (!user->OS()->isDirectory(dir)) {  // does /var exist?
			user->OS()->MakeDirectory(dir,0775);
			user->OS()->changeOwnerGroup(dir,"root","other");
			user->OS()->changePermissions(dir,0775);
		    }
		    dir += "/dt";
		    user->OS()->MakeDirectory(dir,0755);
		    user->OS()->changeOwnerGroup(dir,"root","other");
		    user->OS()->changePermissions(dir,0755);
		}
		dir += "/appconfig";
		user->OS()->MakeDirectory(dir,0755);
		user->OS()->changeOwnerGroup(dir,"bin","bin");
		user->OS()->changePermissions(dir,0755);
	    }
	    dir += "/appmanager";
	    user->OS()->MakeDirectory(dir,0755);
	    user->OS()->changeOwnerGroup(dir,"bin","bin");
	    user->OS()->changePermissions(dir,0755);
	}
	user->OS()->MakeDirectory(dirname_,0755);
	user->OS()->changeOwnerGroup(dirname_,"","");
    }

    // Make /var/dt/tmp directory if not present 

    CString tmp("/var/dt/tmp/");
    if (!user->OS()->isDirectory(tmp)) {   // does tmp exist?
	user->OS()->MakeDirectory(tmp,0777);
	user->OS()->changeOwnerGroup(tmp,"root","other");
	user->OS()->changePermissions(tmp,0777);

        tmp += userhostdir;
        user->OS()->MakeDirectory(tmp, 0755);
        user->OS()->changeOwnerGroup(tmp,"","");
    }
}


/********************************************************************
 * 
 *  TraversePath()
 *
 * 	Parse a given search path, using comma (,) and colon (:) as
 *      delimiters.  Pass each path element to another function.
 ********************************************************************/

void AppManagerDirectory::TraversePath()
{
    if (!appsp_.isNull()) {
	CTokenizedString subpath(appsp_,":");
	CString dir = subpath.next();
	while (!dir.isNull()) {
	    GatherAppsFromASearchElement (dir);
	    dir = subpath.next();
	    if (langVersionFound && dir == "/usr/dt/appconfig/appmanager/C")
		dir = subpath.next();
	}
    }
}

/********************************************************************
 * 
 *  goodFile()
 *
 * 	Ignore '.' and '..' and make sure file doesn't exist.
 *
 ********************************************************************/

int AppManagerDirectory::goodFile
	(
	const CString & path,
	const CString & fname
	) const
{
    if (fname == "." || fname == "..")
	return 0;

    CString filnam(path + "/" + fname);

    if (user_->OS()->FileExists(filnam))
	return 0;

    return 1;
}


/*********************************************************************
 *
 *  GatherAppsFromASearchElement()
 *
 * 	Given search path element (host name or path name), construct
 *      the appropriate path for the various desktop subsystems.
 *
 *      A path is constructed so that each element is appended to
 *      the tail.
 *********************************************************************/
void AppManagerDirectory::GatherAppsFromASearchElement
	(
	const CString & path
	) 
{
    if (path.length()) {
	CString source(path);
	if (source.contains("/%L")) {
	    if (user_->OS()->LANG() != "C") {
		source.replace("%L",user_->OS()->LANG());
		if (!user_->OS()->isDirectory(source))
		     return;
		if (source.contains("/usr/dt/appconfig/appmanager/"))
		    langVersionFound = 1;
	    }
	    else
		return;
	}

	else if (!user_->OS()->isDirectory(source))
	    return;

	DirectoryIterator iter (source);
	struct dirent * direntry;

	user_->OS()->setUserId();

	while (direntry = iter()) {
	    CString dname(direntry->d_name);
	    if (user_->OS()->isDirectory(source + "/" + dname)
	     || user_->OS()->isFile(source + "/" + dname))
		if (goodFile(dirname_, dname))
		    user_->OS()->symbolicLink (source + "/" + dname, 
					       dirname_ + "/" + dname);
	}

	user_->OS()->setUserId();
    }
}

int main (int argc, char **argv)
{
    options = new Options(argc, argv);

    UnixEnvironment * os = new UnixEnvironment;

    CDEEnvironment * user = new CDEEnvironment(0, os);

    LegacyCleanUp(user);

    CString app(getenv("DTAPPSEARCHPATH"));

    AppManagerDirectory * appman = new AppManagerDirectory(user, app);
    appman->TraversePath();

    user->OS()->changePermissions((*appman)(),0555);

    return 0;
}
